Sources/NIO/SocketChannel.swift

```diff
@@ -520,7 +520,9 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
 
         // Was not registered yet so do it now.
         do {
-            try self.safeRegister(interested: .read)
+            // Depending on if we have a readPending or autoRead is one we need to use a different IOEvent to regsiter with the Selector.
+            let ioEvent: IOEvent = readPending || autoRead ? .read : .none
```

Comment:
I think this is semantically wrong. This means that when you `register` a channel, the first `read` skips the pipeline entirely: the pipeline never sees `read`.

That's just wrong, I think. My suspicion is that after `pipeline.fireChannelRegistered0` we should call `readIfNeeded0`, which already knows how to handle `autoRead`. That way the pipeline will actually see a `read`, rather than getting the first read magically for free.

Reply:
@Lukasa so what you suggest is that we always register with `.none` and then call `readIfNeeded0()`after fireChannelRegistered0() correct ? This sounds like a good idea... The only downside is that it will cause one more sys call. But I think that's ok as like you stated calling readIfNeeded0() at least give the user a chance to do something in read.

Reply:
Yeah, I think that's what I'm getting at. I agree the extra syscall is sad, but it fits better with the architecture.

Reply:
@Lukasa good suggestion... Let me do it :)